opm/simulators/flow/TemperatureModel.hpp

```diff
@@ -0,0 +1,503 @@
+            computeStorageTerm(globI, storage);
+            this->energyVector_[globI] += storefac * ( getValue(storage) - storage1_[globI] );
+            if (globI == 0) {
+                //std::cout << "storage " << getValue(storage) << " " << storage1_[globI] << " " << storefac * ( getValue(storage) - storage1_[globI] ) << " " <<  storefac * storage.derivative(Indices::temperatureIdx) << " " << dt/(3600*24) << std::endl;
+            }
+            (*this->energyMatrix_)[globI][globI][0][0] += storefac * storage.derivative(Indices::temperatureIdx);
+        }
+
+        const auto& neighborInfo = simulator_.model().linearizer().getNeighborInfo();
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+        for (unsigned globI = 0; globI < numCells; ++globI) {
+            const auto& nbInfos = neighborInfo[globI];
+            for (const auto& nbInfo : nbInfos) {
+                unsigned globJ = nbInfo.neighbor;
+                assert(globJ != globI);
+
+                // compute convective flux
+                Evaluation flux = 0.0;
+                computeFluxTerm(globI, globJ, nbInfo.res_nbinfo, flux);
+                this->energyVector_[globI] += getValue(flux);
+                (*this->energyMatrix_)[globI][globI][0][0] += flux.derivative(Indices::temperatureIdx);
+                (*this->energyMatrix_)[globJ][globI][0][0] -= flux.derivative(Indices::temperatureIdx);
+
+                // compute conductive flux
+                Evaluation heatFlux = 0.0;
+                computeHeatFluxTerm(globI, globJ, nbInfo.res_nbinfo, heatFlux);
+                this->energyVector_[globI] += getValue(heatFlux);
+                (*this->energyMatrix_)[globI][globI][0][0] += heatFlux.derivative(Indices::temperatureIdx);
+                (*this->energyMatrix_)[globJ][globI][0][0] -= heatFlux.derivative(Indices::temperatureIdx);
+            }
+        }
+
+        // Well terms
+        const auto& wellPtrs = simulator_.problem().wellModel().localNonshutWells();
+        for (const auto& wellPtr : wellPtrs) {
+            this->assembleEquationWell(*wellPtr);
+        }
+
+        // Set dirichlet conditions for overlapping cells
+        if (simulator_.gridView().comm().size() > 1) {
+            //loop over precalculated overlap rows and columns
+            for (const auto row : overlapRows_)
+            {
+                // Zero out row.
+                (*this->energyMatrix_)[row] = 0.0;
+
+                //diagonal block set to diag(1.0).
+                (*this->energyMatrix_)[row][row][0][0] = 1.0;
+            }
+        }
+
+#warning fix this
+        // Comunicate the residual
+        std::vector<EnergyVector> tmp;
+        tmp.resize(1);
+        tmp[0] = this->energyVector_;
+        auto handle = VectorVectorDataHandle<GridView, std::vector<EnergyVector>>(tmp,simulator_.gridView());
+        simulator_.gridView().communicate(handle, Dune::InteriorBorder_All_Interface,Dune::ForwardCommunication);
```

Comment:
@blattms I need a VectorDataHandle not a VectorVectorDataHandle. Do I need to make my own Handle for this or can I use an existing one.

Reply:
there seems to be `GridCommHandleGhostSync` in opm/models/parallel/gridcommhandles.hh that does this, hopefully.

Reply:
Thanks. I did the trick.